{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "\n",
    "#Load in the integrated spleen dataset\n",
    "adata_pl_raw=anndata.read_h5ad('/home/nikvaku/snic2022-6-312/LabMemberScratchDir/Nikhilesh/Inter_data/spleen_merged_raw.h5ad')\n",
    "adata_b_cells=anndata.read_h5ad(\"/home/nikvaku/snic2022-6-312/LabMemberScratchDir/Nikhilesh/Final_Data/spleen_b_cells.h5ad\")\n",
    "adata_t_cells=anndata.read_h5ad(\"/home/nikvaku/snic2022-6-312/LabMemberScratchDir/Nikhilesh/Final_Data/spleen_t_cells.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#GO Analysis\n",
    "import wget\n",
    "import os\n",
    "from goatools import obo_parser\n",
    "go_db_url=' http://purl.obolibrary.org/obo/go/go-basic.obo'\n",
    "data_folder=\"/home/nikvaku/snic2022-6-312/LabMemberScratchDir/Nikhilesh/Raw_data/data\"# Replace this with your data folder\n",
    "\n",
    "if(not os.path.isfile(data_folder+'/go-basic.obo')):\n",
    "    go_obo = wget.download(go_db_url, data_folder+'/go-basic.obo')\n",
    "else:\n",
    "    go_obo = data_folder+'/go-basic.obo'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a id2gos format file using Trianotate\n",
    "triannotate=pd.read_csv('/home/nikvaku/snic2022-6-312/LabMemberScratchDir/Nikhilesh/Raw_data/aPlwal.pri.V2.genome.annots.tsv',sep='\\t')\n",
    "triannotate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creating a GODAG for Pleurodeles\n",
    "\n",
    "from collections import defaultdict\n",
    "\n",
    "eggnog_mapper = defaultdict(list)\n",
    "preferred_names = defaultdict(int)\n",
    "\n",
    "for i, row in triannotate.iterrows():\n",
    "    gene_id = row['#gene_id']\n",
    "    preferred_name = row['EggNM.Preferred_name']\n",
    "    go_terms = row['EggNM.GOs']\n",
    "\n",
    "    if go_terms != '.':\n",
    "        # Determine the key based on the preferred name and gene ID\n",
    "        if preferred_name != '.':\n",
    "            key = preferred_name\n",
    "            # Update the preferred_names dictionary\n",
    "            preferred_names[preferred_name] += 1\n",
    "        else:\n",
    "            key = gene_id\n",
    "            # Check if the gene_id has appeared before\n",
    "            if gene_id in preferred_names:\n",
    "                # Use the preferred name from the first appearance\n",
    "                key = f\"{gene_id}-{preferred_names[gene_id]}\"\n",
    "                preferred_names[gene_id] += 1\n",
    "\n",
    "        # Check if the key already exists in the dictionary\n",
    "        if key in eggnog_mapper:\n",
    "            # Append a suffix starting from -1\n",
    "            suffix = 1\n",
    "            new_key = f\"{key}-{suffix}\"\n",
    "            while new_key in eggnog_mapper:\n",
    "                suffix += 1\n",
    "                new_key = f\"{key}-{suffix}\"\n",
    "            key = new_key\n",
    "\n",
    "        eggnog_mapper[key].append(go_terms)\n",
    "\n",
    "# Convert defaultdict to a regular dictionary\n",
    "eggnog_mapper = dict(eggnog_mapper)\n",
    "\n",
    "#Remove values with only '.' in the values\n",
    "eggnog_mapper_filtered = {}\n",
    "gene_list = []\n",
    "for key in eggnog_mapper.keys():\n",
    "    if eggnog_mapper[key] == '.':\n",
    "        continue\n",
    "    else:\n",
    "        eggnog_mapper_filtered[key] = eggnog_mapper[key]\n",
    "        gene_list.append(key)\n",
    "\n",
    "#Write the mapper to create a id2gos file\n",
    "with open('/home/nikvaku/snic2022-6-312/LabMemberScratchDir/Nikhilesh/Inter_data/id2gos.txt', 'w') as f: #Change this to where you want the dataset\n",
    "    for key, go_terms in eggnog_mapper.items():\n",
    "        # Join the list of GO IDs with semi-colons\n",
    "        go_ids_str = go_terms[0].replace(',', ';') if len(go_terms) == 1 else ';'.join(go_terms)\n",
    "        # Write the gene name, a tab, and the list of GO IDs\n",
    "        f.write(f\"{key}\\t{go_ids_str}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Collecting the top 200 differentially expressed gene\n",
    "\n",
    "#For T cells\n",
    "sc.tl.rank_genes_groups(adata_t_cells,groupby='leiden', method='wilcoxon')\n",
    "sc.pl.rank_genes_groups(adata_t_cells, n_genes=25, sharey=False)\n",
    "cluster0=sc.get.rank_genes_groups_df(adata_t_cells,group='0',key='rank_genes_groups')\n",
    "cluster1=sc.get.rank_genes_groups_df(adata_t_cells,group='1',key='rank_genes_groups')\n",
    "cluster2=sc.get.rank_genes_groups_df(adata_t_cells,group='2',key='rank_genes_groups')\n",
    "cluster3=sc.get.rank_genes_groups_df(adata_t_cells,group='3',key='rank_genes_groups')\n",
    "cluster4=sc.get.rank_genes_groups_df(adata_t_cells,group='4',key='rank_genes_groups')\n",
    "cluster5=sc.get.rank_genes_groups_df(adata_t_cells,group='5',key='rank_genes_groups')\n",
    "cluster6=sc.get.rank_genes_groups_df(adata_t_cells,group='6',key='rank_genes_groups')\n",
    "cluster7=sc.get.rank_genes_groups_df(adata_t_cells,group='7',key='rank_genes_groups')\n",
    "cluster8=sc.get.rank_genes_groups_df(adata_t_cells,group='8',key='rank_genes_groups')\n",
    "cluster0=cluster0['names'][:200]\n",
    "cluster1=cluster1['names'][:200]\n",
    "cluster2=cluster2['names'][:200]\n",
    "cluster3=cluster3['names'][:200]\n",
    "cluster4=cluster4['names'][:200]\n",
    "cluster5=cluster5['names'][:200]\n",
    "cluster6=cluster6['names'][:200]\n",
    "cluster7=cluster7['names'][:200]\n",
    "cluster8=cluster8['names'][:200]\n",
    "\n",
    "#For B cells\n",
    "sc.tl.rank_genes_groups(adata_b_cells,groupby='leiden',method='wilcoxon')\n",
    "sc.pl.rank_genes_groups(adata_b_cells,n_genes=25,sharey=False)\n",
    "cluster0_b=sc.get.rank_genes_groups_df(adata_b_cells,group='0',key='rank_genes_groups')\n",
    "cluster1_b=sc.get.rank_genes_groups_df(adata_b_cells,group='1',key='rank_genes_groups')\n",
    "cluster2_b=sc.get.rank_genes_groups_df(adata_b_cells,group='2',key='rank_genes_groups')\n",
    "cluster3_b=sc.get.rank_genes_groups_df(adata_b_cells,group='3',key='rank_genes_groups')\n",
    "cluster4_b=sc.get.rank_genes_groups_df(adata_b_cells,group='4',key='rank_genes_groups')\n",
    "cluster5_b=sc.get.rank_genes_groups_df(adata_b_cells,group='5',key='rank_genes_groups')\n",
    "cluster6_b=sc.get.rank_genes_groups_df(adata_b_cells,group='6',key='rank_genes_groups')\n",
    "cluster7_b=sc.get.rank_genes_groups_df(adata_b_cells,group='7',key='rank_genes_groups')\n",
    "cluster8_b=sc.get.rank_genes_groups_df(adata_b_cells,group='8',key='rank_genes_groups')\n",
    "cluster9_b=sc.get.rank_genes_groups_df(adata_b_cells,group='9',key='rank_genes_groups')\n",
    "cluster10_b=sc.get.rank_genes_groups_df(adata_b_cells,group='10',key='rank_genes_groups')\n",
    "cluster11_b=sc.get.rank_genes_groups_df(adata_b_cells,group='11',key='rank_genes_groups')\n",
    "cluster12_b=sc.get.rank_genes_groups_df(adata_b_cells,group='12',key='rank_genes_groups')\n",
    "cluster0_b=cluster0_b['names'][:200]\n",
    "cluster1_b=cluster1_b['names'][:200]\n",
    "cluster2_b=cluster2_b['names'][:200]\n",
    "cluster3_b=cluster3_b['names'][:200]\n",
    "cluster4_b=cluster4_b['names'][:200]\n",
    "cluster5_b=cluster5_b['names'][:200]\n",
    "cluster6_b=cluster6_b['names'][:200]\n",
    "cluster7_b=cluster7_b['names'][:200]\n",
    "cluster8_b=cluster8_b['names'][:200]\n",
    "cluster9_b=cluster9_b['names'][:200]\n",
    "cluster10_b=cluster10_b['names'][:200]\n",
    "cluster11_b=cluster11_b['names'][:200]\n",
    "cluster12_b=cluster12_b['names'][:200]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Gene Ontology Enrichment Analysis\n",
    "\n",
    "from goatools.go_enrichment import GOEnrichmentStudy\n",
    "from goatools.anno.idtogos_reader import IdToGosReader\n",
    "import scipy\n",
    "import fisher\n",
    "from goatools.base import download_go_basic_obo\n",
    "def go_enrichment(cluster):\n",
    "    \n",
    "    id2gos_file_path = '/home/nikvaku/snic2022-6-312/LabMemberScratchDir/Nikhilesh/Inter_data/id2gos.txt'\n",
    "\n",
    "    # Path to the Gene Ontology OBO file\n",
    "    obo_file='/home/nikvaku/snic2022-6-312/LabMemberScratchDir/Nikhilesh/Raw_data/data/go-basic.obo'\n",
    "    obo_file =GODag(obo_file)\n",
    "# Create an instance of IdToGosReader and read the gene to GO annotations\n",
    "    id2gos_reader = IdToGosReader(id2gos_file_path)\n",
    "    id2gos_dict = id2gos_reader.get_id2gos()\n",
    "\n",
    "    # Background set of genes (all genes in your study)\n",
    "    background_genes = set(adata_pl_raw.var_names)\n",
    "\n",
    "    # Your set of genes of interest\n",
    "    genes_of_interest = list(cluster)  # Replace with your actual gene IDs\n",
    "\n",
    "    # Create an instance of GOEnrichmentStudy\n",
    "    go_enrichment = GOEnrichmentStudy(\n",
    "        background_genes, # List of all genes in my study\n",
    "        id2gos_dict,# List of Pleuro genes\n",
    "        obo_file,\n",
    "        propagate_counts=False,# Set to True if you want to propagate counts up the GO hierarchy\n",
    "        alpha=0.01,  # Set your desired significance level\n",
    "        methods=['fdr_bh']  # Set the multiple testing correction method\n",
    "    )\n",
    "\n",
    "    # Run the GO enrichment analysis\n",
    "    go_results = go_enrichment.run_study(genes_of_interest)\n",
    "    return go_results\n",
    "\n",
    "#Creating a filtered table for the GO analysis\n",
    "def create_go_table(go_results):\n",
    "    goea_results_sig = [r for r in go_results if r.p_fdr_bh < 0.05 and len(r.study_items) > 1]\n",
    "\n",
    "    # Create DataFrame from the significant results\n",
    "    GO = pd.DataFrame([{\n",
    "        'GO': res.GO,\n",
    "        'term': res.goterm.name,\n",
    "        'class': res.goterm.namespace,\n",
    "        'p': res.p_uncorrected,\n",
    "        'p_corr': res.p_fdr_bh,\n",
    "        'n_genes': res.ratio_in_study[0],\n",
    "        'n_study': res.ratio_in_study[1],\n",
    "        'n_go': len(res.goterm.get_all_parents()) + 1, # Assuming you want to count the term itself plus all parents\n",
    "        'study_genes': list(res.study_items)\n",
    "    } for res in goea_results_sig])\n",
    "    \n",
    "    return GO\n",
    "\n",
    "\n",
    "#Plot the results of GOEA\n",
    "def go_plot(dataframe):\n",
    "    import numpy as np\n",
    "    import matplotlib.pyplot as plt\n",
    "    import seaborn as sns\n",
    "    import pandas as pd\n",
    "    \n",
    "    df = dataframe.copy()\n",
    "    # Convert p-values to a negative log scale to amplify differences\n",
    "    df['-log10(p_corr)'] = -np.log10(df['p_corr'])\n",
    "\n",
    "    # Sort the DataFrame based on the significance level of GO terms\n",
    "    df_sorted = df.sort_values('-log10(p_corr)', ascending=False)\n",
    "\n",
    "    # Select the top N significant terms for plotting\n",
    "    N = 20\n",
    "    df_top = df_sorted.head(N)\n",
    "    \n",
    "    # Normalize the number of genes for color mapping across the full range (blue to red)\n",
    "    gene_counts_norm = (df_top['n_genes'] - df_top['n_genes'].min()) / (df_top['n_genes'].max() - df_top['n_genes'].min())\n",
    "    colors = plt.cm.coolwarm(gene_counts_norm)\n",
    "\n",
    "    # Create the bar plot\n",
    "    plt.figure(figsize=(10, 8))\n",
    "    ax = sns.barplot(x='-log10(p_corr)', y='term', data=df_top,\n",
    "                     palette=colors,\n",
    "                     edgecolor='black')\n",
    "\n",
    "    # Create color bar for the number of genes\n",
    "    sm = plt.cm.ScalarMappable(cmap=\"coolwarm\", norm=plt.Normalize(vmin=df_top['n_genes'].min(), vmax=df_top['n_genes'].max()))\n",
    "    sm.set_array([])\n",
    "    cbar = plt.colorbar(sm, ax=ax)\n",
    "    cbar.set_label('Number of Genes', rotation=270, labelpad=15)\n",
    "\n",
    "    # Set the plot labels and title\n",
    "    plt.xlabel('-log10(Adjusted P-Value)')\n",
    "    plt.ylabel('GO Term')\n",
    "    plt.title('Cluster12 Significant GO Terms by Adjusted P-Value')\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "#Use these functions for the entire process.\n",
    "go_results = go_enrichment(cluster12_b)\n",
    "GO_table = create_go_table(go_results)\n",
    "go_plot(GO_table)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
